### PR TITLE
JAMES-1791 Support inline attachments

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Attachment.java [hunk 1]
```diff
@@ -135,7 +135,7 @@ public long getSize() {
         return cid;
     }
 
-    public boolean isInline() {
+    public boolean isIsInline() {
```
Comment (MichaelBailly) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Attachment.java]: areAreYouSure ?

Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Attachment.java]: Yes, explaination:
Jackson removes the verb from the accessors (get, is ...) in order to producer the JSON.
The JMAP spécification is waiting for "isInline": "..."
So we have to double the 'is' verb.
You can found similar method in the Message class.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 2]
```diff
@@ -120,14 +124,16 @@
     private final UidProvider uidProvider;
     private final CassandraTypesProvider typesProvider;
     private final int maxRetries;
+    private final AttachmentMapper attachmentMapper;
 
-    public CassandraMessageMapper(Session session, UidProvider uidProvider, ModSeqProvider modSeqProvider, MailboxSession mailboxSession, int maxRetries, CassandraTypesProvider typesProvider) {
+    public CassandraMessageMapper(Session session, UidProvider uidProvider, ModSeqProvider modSeqProvider, MailboxSession mailboxSession, int maxRetries, CassandraTypesProvider typesProvider, AttachmentMapper attachmentMapper) {
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I'm not convinced a mapper should access another mapper

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 3]
```diff
@@ -333,13 +339,18 @@ private PropertyBuilder getPropertyBuilder(Row row) {
...
         switch (fetchType) {
         case Full:
         case Body:
-            return row.getList(ATTACHMENTS_IDS, String.class)
+            return row.getList(ATTACHMENTS, UDTValue.class)
                     .stream()
-                    .map(AttachmentId::from)
+                    .map(Throwing.function(x -> 
+                        MessageAttachment.builder()
+                            .attachment(attachmentMapper.getAttachment(AttachmentId.from(x.getString(Attachments.ID))))
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: you are doing n+1 queries, it looks like a bad idea, you should fetch attachment for n messages in a single query

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 4]
```diff
@@ -375,8 +386,12 @@ private MessageMetaData save(Mailbox mailbox, MailboxMessage message) throws Mai
                         .setString(Properties.VALUE, x.getValue()))
                     .collect(Collectors.toList()))
                 .value(TEXTUAL_LINE_COUNT, message.getTextualLineCount())
-                .value(ATTACHMENTS_IDS, message.getAttachmentsIds().stream()
-                    .map(AttachmentId::getId)
+                .value(ATTACHMENTS, message.getAttachments().stream()
+                    .map(x -> typesProvider.getDefinedUserType(ATTACHMENTS)
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: udt deserves it own method

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraMessageModule.java [hunk 5]
```diff
@@ -87,13 +87,19 @@ public CassandraMessageModule() {
...
                 SchemaBuilder.createType(CassandraMessageTable.PROPERTIES)
                     .ifNotExists()
                     .addColumn(CassandraMessageTable.Properties.NAMESPACE, text())
                     .addColumn(CassandraMessageTable.Properties.NAME, text())
-                    .addColumn(CassandraMessageTable.Properties.VALUE, text())));
+                    .addColumn(CassandraMessageTable.Properties.VALUE, text())),
+            new CassandraType(CassandraMessageTable.ATTACHMENTS,
+                    SchemaBuilder.createType(CassandraMessageTable.ATTACHMENTS)
+                        .ifNotExists()
+                        .addColumn(CassandraMessageTable.Attachments.ID, text())
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraMessageModule.java]: no name ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/MessageAttachment.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+
+        private Attachment attachment;
+        private Optional<String> cid;
+        private boolean isInline;
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/MessageAttachment.java]: I really prefer Boolean and a default value in builder

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/MessageAttachment.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            return this;
+        }
+
+        public Builder cid(String cid) {
+            this.cid = Optional.fromNullable(cid);
+            return this;
+        }
+
+        public Builder isInline(boolean isInline) {
+            this.isInline = isInline;
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/MessageAttachment.java]: I guess isInline requires cid ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 8]
```diff
@@ -60,25 +69,51 @@
...
+                .build();
+    }
+
+    private Optional<ContentTypeField> getContentTypeField(Entity entity) {
+        return castField(entity.getHeader().getField(CONTENT_TYPE), ContentTypeField.class);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <U extends ParsedField> Optional<U> castField(Field field, Class<U> clazz) {
+        if (field == null || !clazz.isInstance(field)) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: I would like the type mismatch to throw

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java [hunk 9]
```diff
@@ -742,11 +774,15 @@ private MailboxMessage retrieveMessageFromStorage(MailboxMessage message) throws
...
+    private SimpleMailboxMessage createMessage(Mailbox mailbox, String content, int bodyStart, PropertyBuilder propertyBuilder, List<MessageAttachment> attachments) {
+        return new SimpleMailboxMessage(new Date(), content.length(), bodyStart, new SharedByteArrayInputStream(content.getBytes()), new Flags(), propertyBuilder, mailbox.getMailboxId(), attachments);
     }
     
     private SimpleMailboxMessage createMessage(Mailbox mailbox, String content, int bodyStart, PropertyBuilder propertyBuilder) {
         return new SimpleMailboxMessage(new Date(), content.length(), bodyStart, new SharedByteArrayInputStream(content.getBytes()), new Flags(), propertyBuilder, mailbox.getMailboxId());
     }
+
+    private void storeAttachment(Attachment attachment) throws MailboxException {
+        attachmentMapper.storeAttachment(attachment);
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: it's so simple I wouldn't extract it

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 10]
```diff
@@ -317,8 +318,9 @@ public void assertSecondAttachment(DataTable attachmentProperties) throws Throwa
     }
 
     private void assertAttachment(String attachment, DataTable attachmentProperties) {
-        attachmentProperties.asMap(String.class, String.class).entrySet()
+        attachmentProperties.asList(TableRow.class)
             .stream()
-            .map(entry -> response.body(attachment + "." + entry.getKey(), equalTo(entry.getValue())));
+            .peek(System.out::println)
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: debug detected

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 2 above)
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I hesitate on that, then ask Raphael; we don't see any problem as mappers are on the same layer.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 3 above)
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: OK

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 4 above)
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: OK

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/MessageAttachment.java (hunk 6 above)
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/MessageAttachment.java]: OK

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/MessageAttachment.java (hunk 7 above)
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/MessageAttachment.java]: Right

File: mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMailboxMessage.java [hunk 11]
```diff
@@ -355,7 +355,7 @@ public int compareTo(MailboxMessage other) {
     }
 
     @Override
-    public List<AttachmentId> getAttachmentsIds() {
+    public List<MessageAttachment> getAttachments() {
         throw new NotImplementedException("Attachments Ids not implemented");
```
Comment (rouazana) [inline mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMailboxMessage.java]: Please adapt the error message

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 8 above)
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: Should we fail when an header is invalid, i'm not sure.

File: mailbox/jcr/src/main/java/org/apache/james/mailbox/jcr/mail/model/JCRMailboxMessage.java [hunk 12]
```diff
@@ -680,7 +680,7 @@ public int compareTo(MailboxMessage other) {
     }
 
     @Override
-    public List<AttachmentId> getAttachmentsIds() {
+    public List<MessageAttachment> getAttachments() {
         throw new NotImplementedException("Attachments Ids not implemented");
```
Comment (rouazana) [inline mailbox/jcr/src/main/java/org/apache/james/mailbox/jcr/mail/model/JCRMailboxMessage.java]: here too

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java [hunk 13]
```diff
@@ -577,7 +577,7 @@ public String toString() {
     
 
     @Override
-    public List<AttachmentId> getAttachmentsIds() {
+    public List<MessageAttachment> getAttachments() {
         throw new NotImplementedException("Attachments Ids not implemented");
```
Comment (rouazana) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: idem

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java (hunk 9 above)
Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: OK

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/model/MaildirMessage.java [hunk 14]
```diff
@@ -277,7 +278,7 @@ public InputStream getHeaderContent() throws IOException {
     }
 
     @Override
-    public List<AttachmentId> getAttachmentsIds() {
+    public List<MessageAttachment> getAttachments() {
         throw new NotImplementedException("Attachments Ids not implemented");
```
Comment (rouazana) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/model/MaildirMessage.java]: idem

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java (hunk 10 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: Caught

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 15]
```diff
@@ -60,25 +69,51 @@
...
-                Optional<String> contentType = contentType(contentTypeField);
-                Optional<String> name = name(contentTypeField);
-                
-                attachments.add(Attachment.builder()
-                        .bytes(getBytes(messageWriter, entity.getBody()))
-                        .type(contentType.or(DEFAULT_CONTENT_TYPE))
-                        .name(name)
-                        .build());
+            if (entity.isMultipart()) {
+                attachments.addAll(listAttachments((Multipart) entity.getBody()));
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: Unfortunately entity.isMultipart doesn't check that the type of the body is Multipart. Can you make an explicit check before casting?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMappersTest.java [hunk 16]
```diff
@@ -30,7 +30,7 @@
 
 @RunWith(ContractSuite.class)
 @ContractImpl(CassandraMapperProvider.class)
-public class CassandraMappersTests {
+public class CassandraMappersTest {
```
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMappersTest.java]: Should be in previous commit?

**Comment by rouazana** [comment]
Can you check how oneAttachmentAndSomeTextInlined.eml is displayed via JMAP? I would like to be sure the text is not lost.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraMessageModule.java (hunk 5 above)
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraMessageModule.java]: OK in a separate commit.

File: mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMailboxMessage.java (hunk 11 above)
Comment (aduprat) [inline mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMailboxMessage.java]: OK

File: mailbox/jcr/src/main/java/org/apache/james/mailbox/jcr/mail/model/JCRMailboxMessage.java (hunk 12 above)
Comment (aduprat) [inline mailbox/jcr/src/main/java/org/apache/james/mailbox/jcr/mail/model/JCRMailboxMessage.java]: OK

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java (hunk 13 above)
Comment (aduprat) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: OK

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/model/MaildirMessage.java (hunk 14 above)
Comment (aduprat) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/model/MaildirMessage.java]: OK

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMappersTest.java (hunk 16 above)
Comment (aduprat) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMappersTest.java]: Previous commit was for fixing Memory backend

**Comment by aduprat** [comment]
Will do in next issue:
"Attachment : Support messages with mixed content"

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 15 above)
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: OK

**Comment by rouazana** [comment]
:+1:

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 17]
```diff
@@ -333,19 +342,45 @@ private PropertyBuilder getPropertyBuilder(Row row) {
...
                     .collect(ImmutableCollectors.toImmutableList());
         default:
             return ImmutableList.of();
         }
     }
 
+    private Map<AttachmentId,Attachment> attachmentsById(Row row, List<UDTValue> udtValues) {
+        Map<AttachmentId, Attachment> map = new HashMap<>();
+        attachmentMapper.getAttachments(attachmentIds(udtValues)).stream()
+            .forEach(att -> map.put(att.getAttachmentId(), att));
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: you can collect to a map to achieve what you want

**Comment by mbaechler** [comment]
what about documentation update ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 17 above)
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: OK

**Comment by aduprat** [comment]
Ok

**Comment by rouazana** [comment]
:+1:

**Comment by mbaechler** [comment]
:+1:

**Comment by rouazana** [comment]
:+1:

**Comment by aduprat** [comment]
test this please

**Pull request merged**